### PR TITLE
Refactor bootstrap retrieval in `HttpChannelPool`

### LINKED ISSUE #5129: Refactor Bootstrap retrieval in `HttpChannelPool`
We now have fairly complex logic for creating and getting `Bootstrap`s in `HttpChannelPool`.
It's 3-dimensional access that depend on:
- inet or unix
- session protocol; and
- WebSocket or not

We need to extract this data structure and lookup logic into its own class to simplify it.
```
final class Bootstraps {

  Bootstraps() {...} // Initialize data structure whose key is composed of the above 3 dimensions.

  Bootstrap get(SocketAddress remoteAddress, SessionProtocol desiredProtocol,
          SerializationFormat serializationFormat) {...}
}
```

### PR DESCRIPTION
**Pull request by russell-yoo**
Related: #5129

Motivation:
The `bootstrap` creation logic in the `HttpChannelPool`  is fairly complex. We want to simplify it by making it a separate class and data structure. from @minwoox 

Modifications:
- `inetBootstraps` and `unixBootstraps` were fields in `HttpChannelPool`, but have now been changed to fields in the own class `Bootstraps`.

Result:
- Closes #5129. (If this resolves the issue.)
- We've separate the bootstrap retrieval logic from `HttpChannelPool` to make it simple.

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/client/Bootstraps.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+            unixBootstraps = null;
+        }
+        connectTimeoutMillis = (Integer) inetBaseBootstrap.config().options()
+                                                          .get(ChannelOption.CONNECT_TIMEOUT_MILLIS);
+    }
+
+    Bootstrap get(SocketAddress remoteAddress, SessionProtocol desiredProtocol,
+                  SerializationFormat serializationFormat) {
+        if (!httpAndHttpsValues().contains(desiredProtocol)) {
+            throw new NullPointerException("Unsupported session protocol: " + desiredProtocol);
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/Bootstraps.java:71]: `IllegalArgumentException`?

File: core/src/main/java/com/linecorp/armeria/client/Bootstraps.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        }
+
+        return select(unixBootstraps, desiredProtocol, serializationFormat);
+    }
+
+    int getConnectTimeoutMillis() {
+        return connectTimeoutMillis;
+    }
+
+    private Bootstrap select(Bootstrap[][] bootstraps, SessionProtocol desiredProtocol, SerializationFormat serializationFormat) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/Bootstraps.java:92]: we can make this method, `setBootstrap`, and `toIndex` methods static.

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/Bootstraps.java:92]: Please, also check the checkstyle. A line must not exceed 112 length. 😉

File: core/src/main/java/com/linecorp/armeria/client/Bootstraps.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+    private int toIndex(SerializationFormat serializationFormat) {
+        return toIndex(serializationFormat == SerializationFormat.WS);
+    }
+
+    private SslContext determineSslContext(SessionProtocol desiredProtocol) {
+        return desiredProtocol == SessionProtocol.H1 || desiredProtocol == SessionProtocol.H1C ?
+               sslCtxHttp1Only : sslCtxHttp1Or2;
+    }
+}
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/Bootstraps.java:136]: nit: missing new line.

**Review by minwoox — COMMENTED** [review]
Thanks for the PR. 😉

File: core/src/main/java/com/linecorp/armeria/client/Bootstraps.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+    private int toIndex(boolean webSocket) {
+        return webSocket ? 1 : 0;
+    }
+
+    private int toIndex(SerializationFormat serializationFormat) {
+        return toIndex(serializationFormat == SerializationFormat.WS);
+    }
+
+    private SslContext determineSslContext(SessionProtocol desiredProtocol) {
```
Comment (jrhee17) [inline core/src/main/java/com/linecorp/armeria/client/Bootstraps.java:132]: Do you mind deduping this with `HttpChannelPool#determineSslContext`? I think `HttpChannelPool` can call `Bootstraps#determineSslContext`

File: core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java [hunk 5]
```diff
@@ -114,55 +107,7 @@ final class HttpChannelPool implements AsyncCloseable {
         this.sslCtxHttp1Only = sslCtxHttp1Only;
         this.sslCtxHttp1Or2 = sslCtxHttp1Or2;
```
Comment (jrhee17) [inline core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java:108]: nit; I think we may also be able to remove this if my other comment is handled

**Review by jrhee17 — COMMENTED** [review]
Looks good overall! Left some minor suggestions

File: core/src/main/java/com/linecorp/armeria/client/Bootstraps.java (hunk 1 above)
Comment (russell-yoo) [reply core/src/main/java/com/linecorp/armeria/client/Bootstraps.java:71]: Oh good, I forgot to change the exception I set NPE to something else. 😂 Thanks

File: core/src/main/java/com/linecorp/armeria/client/Bootstraps.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+        if (unixBootstraps == null) {
+            throw new IllegalArgumentException("Domain sockets are not supported by " +
+                                               eventLoop.getClass().getName());
+        }
+
+        return select(unixBootstraps, desiredProtocol, serializationFormat);
+    }
+
+    int getConnectTimeoutMillis() {
```
[the remark this answers is not in the document]
Comment (russell-yoo) [reply core/src/main/java/com/linecorp/armeria/client/Bootstraps.java:88]: The `connectTimeoutMills` came from `inetBaseBootstrap.config`, so I added that value to `Bootstraps`. If I wanted to get that value from the `HttpChannelPool`, I would have to create a new `inetBootstraps` in `cleintFactory`, so that's how I did it. But conceptually, having it in the `httpChannelPool` looks good!!!  😄

[the remark this answers is not in the document]
Comment (russell-yoo) [reply core/src/main/java/com/linecorp/armeria/client/Bootstraps.java:88]: What do you think about taking `clientFacotry`, `inetBootstraps`, and `unixBootstraps` as arguments to the constructor of `Bootstraps`?

This would solve the problem I was having with having to create duplicates.

File: core/src/main/java/com/linecorp/armeria/client/Bootstraps.java (hunk 4 above)
Comment (russell-yoo) [reply core/src/main/java/com/linecorp/armeria/client/Bootstraps.java:132]: That's a good idea. I think using `Bootstraps#determineSslContext` in `HttpConnectionPool` would reduce the fields needed for that method, and also reduce the dedeuping code 😍. Thanks.

File: core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java [hunk 7]
```diff
@@ -945,4 +878,5 @@ private void handlePendingPiggybacks(@Nullable PooledChannel value) {
             }
         }
     }
+
```
Comment (jrhee17) [inline core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java:881]: nit; unnecessary newline

File: core/src/main/java/com/linecorp/armeria/client/Bootstraps.java (hunk 3 above)
Comment (jrhee17) [reply core/src/main/java/com/linecorp/armeria/client/Bootstraps.java:134]: It seems like this comment isn't addressed

File: core/src/main/java/com/linecorp/armeria/client/Bootstraps.java [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+final class Bootstraps {
+    private final Bootstrap[][] inetBootstraps;
+    @Nullable
+    private final Bootstrap[][] unixBootstraps;
+    private final EventLoop eventLoop;
+    private final SslContext sslCtxHttp1Only;
+    private final SslContext sslCtxHttp1Or2;
+
+    Bootstraps(HttpClientFactory clientFactory, EventLoop eventLoop, SslContext sslCtxHttp1Or2,
+               SslContext sslCtxHttp1Only, Bootstrap inetBaseBootstrap, Bootstrap unixBaseBootstrap) {
```
Comment (jrhee17) [inline core/src/main/java/com/linecorp/armeria/client/Bootstraps.java:46]:
```suggestion
               SslContext sslCtxHttp1Only, Bootstrap inetBaseBootstrap, @Nullable Bootstrap unixBaseBootstrap) {
```

**Review by jrhee17 — APPROVED** [review]
Left some minor nits, but looks good to me 👍  Thanks! @russell-yoo 🙇 👍 🚀

File: core/src/main/java/com/linecorp/armeria/client/Bootstraps.java [hunk 9]
```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2020 LINE Corporation
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/Bootstraps.java:2]: nit:
```suggestion
 * Copyright 2023 LINE Corporation
```

File: core/src/main/java/com/linecorp/armeria/client/Bootstraps.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/Bootstraps.java]: How about using `clientFactory` options so that we don't have to create `inetBaseBootstrap` in `HttpChannelPool`?
```java
HttpChannelPool(HttpClientFactory clientFactory, EventLoop eventLoop,
                SslContext sslCtxHttp1Or2, SslContext sslCtxHttp1Only,
                ConnectionPoolListener listener) {
    ...
    this.listener = listener;

    connectTimeoutMillis =
            (Integer) clientFactory.options().channelOptions().get(ChannelOption.CONNECT_TIMEOUT_MILLIS);

    bootstraps = new Bootstraps(clientFactory, eventLoop, sslCtxHttp1Or2, sslCtxHttp1Only);
}

Bootstraps(HttpClientFactory clientFactory, EventLoop eventLoop, SslContext sslCtxHttp1Or2,
           SslContext sslCtxHttp1Only) {...}
```

File: core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java [hunk 10]
```diff
@@ -573,7 +506,7 @@ private void notifyConnect(SessionProtocol desiredProtocol,
                     // Clean up old unhealthy channels by iterating from the beginning of the queue.
                     final Deque<PooledChannel> queue = getPool(protocol, key);
                     if (queue != null) {
-                        for (;;) {
+                        while (true) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java:509]: Could you revert this change?

**Review by minwoox — COMMENTED** [review]
Looks great and left small suggestions. 😉

File: core/src/main/java/com/linecorp/armeria/client/Bootstraps.java (hunk 9 above)
Comment (russell-yoo) [reply core/src/main/java/com/linecorp/armeria/client/Bootstraps.java]: thx

File: core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java (hunk 10 above)
Comment (russell-yoo) [reply core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java]: Ok, sure. I will revert this change

File: core/src/main/java/com/linecorp/armeria/client/Bootstraps.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/Bootstraps.java]: This isn't resolved. 😉

[the remark this answers is not in the document]
Comment (russell-yoo) [reply core/src/main/java/com/linecorp/armeria/client/Bootstraps.java]: I missed it. Thanks @minwoox. I didn't know that I could get `connectTimeoutMills` by `clientFactor.options()` . I'll apply that and post back.

**Review by minwoox — APPROVED** [review]
Thanks, @russell-yoo! 😄

**Review by ikhoon — APPROVED** [review]
Thanks, @russell-yoo! 🚀❤️
